.github/workflows/ci.yaml

```diff
@@ -406,3 +406,28 @@ jobs:
         with:
           name: release-images
           path: ./release-images
+
+  trivy:
+    runs-on: ubuntu-18.04
+    needs: prebuild
+    steps:
+      - name: Checkout code
+      - uses: actions/checkout@v2
+
+      - name: Build an image from Dockerfile
+        run: |
+          docker build -t codercom/code-server:${{ github.sha }} ci/release-image/Dockerfile
```

Comment:
why is building the image needed if we're just doing a filesystem scan?

alternatively, maybe we could also do an image scan? because the images contain other stuff (e.g. OS dependencies) that would not be caught in a filesystem scan

Reply:
I almost messaged you to ask which we should do. 

What would you recommend?

Reply:
I'm not going to make specific _recommendations_ because you have more context about the code-server project than me, but will just share some experience and ideas for you to think about, not in any particular order:

* In the monorepo, we do both a filesystem scan and image scan.
  * The filesystem scan is nice because it runs quickly and can give feedback rapidly, and we can do it on every PR/merge to master without adding much time to the build duration. The downside is that you will either need to ignore errors for the check using `continue-on-error`, or else accept that sometimes the build will start failing until you have a chance to remediate the failing checks. If using `continue-on-error`, you'll need to remember to manually check the tab to see if things were OK, or look out for warnings that GitHub shows (it will show that there was a failure as a warning, but it won't appear red).
  * The image scan is useful because it can detect problems with installed packages (via `apt-get`) which source code scans cannot detect, and ultimately we're responsible for the full supply chain (both our code, third-party dependencies we import with package.json or submodules, upstream container images that we build `FROM`, and packages that we install via `apt-get` or similar)
* In the monorepo, we put our security scanning in a separate workflow and run it on a scheduled basis (nightly).
  * This means that it's not part of the critical path and won't impact PRs. This may be mitigated depending on the test parallelism we're using, if the other builds take longer than these scans and everything happens in parallel, then these checks would not be in the critical path so it's a moot point.
  * We have a central scan workflow that handles both filesystem and image scans. We do this both on commits to our default branch as well as on a nightly basis, this is to catch problems that show up even if we aren't changing the code (i.e. a security issue is discovered in a package we installed with apt-get, so even though we didn't modify any source code, we suddenly went from clean to vulnerable).
  * I broke the workflow so it was convenient that I could turn it off without affecting other stuff

Other thoughts/ideas:

* In the future, we might want to add other scan tools as well (such as Anchore Grype, Twistlock, OpenSCAP) and having these in a separate workflow means they're outside the critical path and won't slow down your normal development. The downside of the separate workflow is that it may take a day to discover vulnerabilities, but I don't think that's a big problem as long as we fix the issues quickly.
* Generating a Software Bill of Materials (i.e. the full list of "ingredients" of a code-server image, including system packages, third-party dependencies brought in via package.json, etc) might be useful in a similar step. This can be done using tools like [Anchore Syft](https://github.com/anchore/syft).
* We may want to upload reports in a different way (e.g. an HTML/CSS report page) to an S3 bucket or something, just to make these results easier for us to process.

I really like the integration with CodeQL that you discovered! Looks like it'll help make things easier to navigate, regardless of how we implement the workflow :)

Reply:
Wow, thank you for the detailed explanation! I think based on all of that, I'm going to make some significant changes:
- move to a separate workflow
- add `continue-on-error` so it doesn't fail PR checks
- scan both repo and image

I'll make those then re-request your reivew!